libr/util/graph.c

```diff
@@ -162,6 +162,16 @@ R_API RGraphNode *r_graph_add_node(RGraph *t, void *data) {
 	return n;
 }
 
+R_API RGraphNode *r_graph_add_nodef(RGraph *graph, void *data, RListFree user_free) {
+	RGraphNode *node = r_graph_add_node (graph, data);
+	if (node) {
+		node->free = user_free;
+	} else if (user_free) {
+		user_free (data);
```

Comment:
I'm not sure about this. If the allocation goes wrong, `data` should still be owned by the caller, no?
The caller would do something like:
```C
n = r_graph_add_nodef (g, data, ufree);
if (!n) {
   ufree (data);
   // free other stuff defined in this function
}
```

Reply:
Yes I was having some doubts on which of the two strategies to choose. Here are the reasons I chose this strategy:
* "This function argument takes ownership, ALWAYS!" seemed like a simpler rule than "It takes ownership but sometimes it doesn't"
* Less error handling code and cleaner code in function using the API to build a graph.
* Handling of the case when construction consists of two steps and first step succeeds transfering the ownership, but second step fails. Here is an example to make the problem clearer:
```c
foo_bar *make_foobar(char *a)
{
    foo * foo_obj = make_foo(a);
    if (!foo_obj) {
       return NULL;
   }
   foo_x(foo_obj);
   foo_bar * foo_bar_obj = malloc(...);
   if (!foo_bar) {
       // What to do now? Ownership of a was already transfered to foo_obj. Can't call the normal destructor of foo, because that would also free a. What if foo_x or even make_foo can modify or replace `a`?
   }
   return foo_bar_obj;
}
```

It seems that neither of approaches is widely used and in most case the approach you suggested means that no one in case of erorr memory is leaked because caller also doesn't release it.  If you consider that my approach is likely to cause confusion I would prefer the third option of not doing the ownership transfer at all and doing a copy when necessary. That way at least there is less conditional code on most levels.

Reply:
Probably both approaches have their own pros/cons, but so far I think we are used to the approach "take ownership only if there's no error". At least, this is what I think is used in r_list and other data structures. @thestr4ng3r @trufae opinions?

Reply:
BTW, see also https://github.com/radareorg/radare2/pull/17621/files#diff-635a23ca710c8cf67d67f28d75cdcf86R24 . I implemented `ht_`, so I'm not sure how much that qualifies as a valid example, but even there, if `n` is not correctly inserted in the ht, it is still freed by the caller.

Reply:
You said this reply was not required any more. So, is this going away? Are changes still missing from this PR?

Reply:
Missed that one due to it being separate file. Removed.